core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -202,6 +212,14 @@ public long acquireExecutionMemory(long required, MemoryConsumer requestingConsu
         logger.debug("Task {} acquired {} for {}", taskAttemptId, Utils.bytesToString(got),
           requestingConsumer);
       }
+
+      if (mode == MemoryMode.OFF_HEAP) {
+        peakOffHeapMemory = Math.max(peakOffHeapMemory,
+          memoryManager.getOffHeapExecutionMemoryUsageForTask(taskAttemptId));
```

Comment:
This introduces the extra lock synchronization to the underlying memory pool. I wonder if we could do a math calculation for the latest peak memory like this:  peakMemory - releasedMemory + gotMemory.

Reply:
Theoretically we can maintain both currentMem and peakMem within `TaskMemoryManager` so that we don’t need to ask `memoryManager`, but on the other hand `memoryManager` is designed to maintain per-task mem usage so by doing this we kinda maintain this in two places. @JoshRosen WDYT

Reply:
Although it's true that there might a bit of redundancy in counting in both places, it seems like there may be reasonable performance justifications for introducing such redundancy.

I don't think it will end up being that much additional code:

Within TaskMemoryManager, I think we'd just need to add a pair of `long` counter fields, one for on-heap and another for off-heap,` then increment them in `acquireExecutionMemory` and decrement them in `releaseExecutionMemory` (since those are narrow waists).

Maybe we should give that a try and see how much net code it ends up adding?

Reply:
I noticed that `releaseExecutionMemory` is not locked, so we need to synchronize on these counters. But I suppose that would be better than synchronization on `MemoryManager` granularity.